storage-bigtable/src/access_token.rs

```diff
@@ -91,41 +105,55 @@ impl AccessToken {
     }
 
     /// Call this function regularly to ensure the access token does not expire
-    pub async fn refresh(&self) {
+    pub fn try_refresh(&self) {
         // Check if it's time to try a token refresh
-        {
-            let token_r = self.token.read().unwrap();
-            if token_r.1.elapsed().as_secs() < token_r.0.expires_in() as u64 / 2 {
-                return;
+        match self.token.read() {
+            Ok(token_r) => {
+                if token_r.1.elapsed().as_secs() < token_r.0.expires_in() as u64 / 2 {
+                    debug!("token is not expired yet");
+                    return;
+                }
+                drop(token_r);
             }
-
-            #[allow(deprecated)]
-            if self
-                .refresh_active
-                .compare_and_swap(false, true, Ordering::Relaxed)
-            {
-                // Refresh already pending
+            Err(_poisoned) => {
+                error!("token is poisoned (read op)");
                 return;
             }
         }
 
-        info!("Refreshing token");
-        match time::timeout(
-            time::Duration::from_secs(5),
-            Self::get_token(&self.credentials, &self.scope),
-        )
-        .await
-        {
-            Ok(new_token) => match (new_token, self.token.write()) {
-                (Ok(new_token), Ok(mut token_w)) => *token_w = new_token,
-                (Ok(_new_token), Err(err)) => warn!("{}", err),
-                (Err(err), _) => warn!("{}", err),
-            },
-            Err(_) => {
-                warn!("Token refresh timeout")
-            }
+        // Refresh already is progress
+        let refresh_progress =
+            self.refresh_active
+                .compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed);
+        if refresh_progress.is_err() {
+            debug!("token update is already in progress");
+            return;
```

Comment:
This change would be good as a separate commit as well.

Reply:
Revert changes with `AccessTokenInner`, please check @CriesofCarrots